middle_end/flambda2/numbers/vector_types.ml

```diff
@@ -60,56 +60,110 @@ end
 
 module Vec128 = struct
   type t =
+    | Unknown128
     | Int8x16
     | Int16x8
     | Int32x4
     | Int64x2
     | Float32x4
     | Float64x2
+    | Any128
 
   let name = function
+    | Unknown128 -> "Unknown128"
     | Int8x16 -> "Int8x16"
     | Int16x8 -> "Int16x8"
     | Int32x4 -> "Int32x4"
     | Int64x2 -> "Int64x2"
     | Float32x4 -> "Float32x4"
     | Float64x2 -> "Float64x2"
+    | Any128 -> "Any128"
 
   let name_lowercase = function
+    | Unknown128 -> "unknown128"
     | Int8x16 -> "int8x16"
     | Int16x8 -> "int16x8"
     | Int32x4 -> "int32x4"
     | Int64x2 -> "int64x2"
     | Float32x4 -> "float32x4"
     | Float64x2 -> "float64x2"
+    | Any128 -> "any128"
+
+  let join l r =
+    (* Least upper bound *)
+    match l, r with
+    | Unknown128, _ -> Unknown128
+    | _, Unknown128 -> Unknown128
+    | Any128, _ -> r
+    | _, Any128 -> l
+    | Int8x16, Int8x16 -> Int8x16
+    | Int8x16, _ -> Unknown128
+    | Int16x8, Int16x8 -> Int16x8
+    | Int16x8, _ -> Unknown128
+    | Int32x4, Int32x4 -> Int32x4
+    | Int32x4, _ -> Unknown128
+    | Int64x2, Int64x2 -> Int64x2
+    | Int64x2, _ -> Unknown128
+    | Float32x4, Float32x4 -> Float32x4
+    | Float32x4, _ -> Unknown128
+    | Float64x2, Float64x2 -> Float64x2
+    | Float64x2, _ -> Unknown128
```

Comment:
Style-wise, the rest of the code usually lists all identical cases, and then a single `(Int8x16 | Inx16x8 | ...), _` at the end to make sure the pattern matching is not fragile.